src/witness/yamlWitness.ml

```diff
@@ -117,3 +154,195 @@ struct
+        ; presub = (fun _ -> raise Not_found)
+        ; spawn  = (fun v d    -> failwith "Cannot \"spawn\" in witness context.")
+        ; split  = (fun d es   -> failwith "Cannot \"split\" in witness context.")
+        ; sideg  = (fun v g    -> failwith "Cannot \"sideg\" in witness context.")
+        }
+      in
+      Spec.query ctx
+    in
+
+    let yaml = Yaml_unix.of_file_exn (Fpath.v (GobConfig.get_string "witness.yaml.validate")) in
+    let yaml_entries = match yaml with
+      | `A yaml_entries -> yaml_entries
+      | _ -> failwith "invalid YAML"
+    in
+
+    let yaml_entries' = List.fold_left (fun yaml_entries' yaml_entry ->
+        let yaml_metadata = Yaml.Util.(yaml_entry |> find_exn "metadata" |> Option.get) in
+        let uuid = Yaml.Util.(yaml_metadata |> find_exn "uuid" |> Option.get |> to_string_exn) in
+        let yaml_location = Yaml.Util.(yaml_entry |> find_exn "location" |> Option.get) in
+        let file = Yaml.Util.(yaml_location |> find_exn "file_name" |> Option.get |> to_string_exn) in
+        let line = Yaml.Util.(yaml_location |> find_exn "line" |> Option.get |> to_float_exn |> int_of_float) in
+        let column = Yaml.Util.(yaml_location |> find_exn "column" |> Option.get |> to_float_exn |> int_of_float) + 1 in
+        let inv = Yaml.Util.(yaml_entry |> find_exn "loop_invariant" |> Option.get |> find_exn "string" |> Option.get |> to_string_exn) in
+        let loc: Cil.location = {
+          file;
+          line;
+          column;
+          byte = -1;
+          endLine = -1;
+          endColumn = -1;
+          endByte = -1;
+        }
+        in
+
+        match Frontc.parse_standalone_exp inv with
+        | inv_cabs ->
+
+          let lvars_opt: LvarS.t option =
+            let (let*) = Option.bind in (* TODO: move to general library *)
+            let* loc_lvars = FileH.find_option file_loc_lvars loc.file in
+            (* for each file, locations (of lvar nodes) have total order, so LocM essentially does binary search *)
+            let* (_, lvars) = LocM.find_first_opt (fun loc' ->
+                CilType.Location.compare loc loc' <= 0 (* allow inexact match *)
+              ) loc_lvars
+            in
+            if LvarS.is_empty lvars then
+              None
+            else
+              Some lvars
+          in
+
+          begin match lvars_opt with
+            | Some lvars ->
+              let module VR = ValidationResult in
+
+              let result = LvarS.fold (fun ((n, _) as lvar) (acc: VR.t) ->
+                  let d = LHT.find lh lvar in
+                  let fd = Node.find_fundec n in
+                  let vars = fd.sformals @ fd.slocals @ global_vars in
+
+                  let genv = Cabs2cil.genvironment in
+                  let env = Hashtbl.copy genv in
+                  List.iter (fun (v: Cil.varinfo) ->
+                      Hashtbl.replace env v.vname (Cabs2cil.EnvVar v, v.vdecl)
+                    ) (fd.sformals @ fd.slocals);
```

Comment:
Did you test with locals that got renamed because several locals of the same name (and different scopes) existed in the function?

Reply:
Good point, I didn't think about it here, but I think there's not much to do about it either. On the witness generation side, there's `InvariantCil.exp_replace_original_name` that maps things to original names. But here for parsing we'd somehow need to know, which of the variables with the same original name was in scope at a particular `location` inside the function _after_ they've been all pulled up and renamed.

As I mentioned in https://github.com/goblint/cil/pull/97, the normal `Cabs2cil` uses tons of global variables to keep track of things and these block scopes and alpha renaming tables (and undos in alpha renaming tables) are among those. But all those structures are mutated, so we cannot go back to a particular location and have the exact state of those internal globals that normally keep track of this. One way would be to keep the CABS of all functions around and redo `Cabs2cil` for the function where the invariant belongs to, so it reconstructs that state, such that we can convert the invariant expression exactly as it would have been at a particular location and abort `Cabs2cil`, but that seems super ugly.

But here's a maybe radical idea: we make CIL not do those renames at all and just continue with a single `fundec` containing multiple local `varinfo`s with same `vname`. Since we identify variables by `vid`, not `vname`, this should have no effect on the semantics, only results presentation (which is nicer with less renaming anyway). And somehow also eliminate the pulling up of all locals from all block scopes, such that the scopes are also later observable.

Anyway, tackling that whole issue is probably too much to do right here, but for the time being, a variable name in the expression would just find the first `varinfo` that has it unrenamed. All the others would have names with suffixes, which won't be used unless you directly also use that name in the invariant, but that would violate the semantics of it being insertable as an assertion.

Reply:
I extracted this problem to #747 for further discussion and future fixes, so this can be merged.